src/solvers/lowering/byte_operators.cpp

```diff
@@ -19,6 +19,168 @@ Author: Daniel Kroening, kroening@kroening.com
+        comp.type(),
+        ns));
+
+      if(component_bits.has_value())
+        offset += bits;
+    }
+
+    return struct_exprt(std::move(operands), target_type);
+  }
+  else if(target_type.id() == ID_array)
+  {
+    const array_typet &array_type = to_array_type(target_type);
+    auto num_elements = numeric_cast<std::size_t>(array_type.size());
+    auto subtype_bits = pointer_offset_bits(array_type.subtype(), ns);
+
+    const std::size_t total_width =
+      to_bitvector_type(bitvector_expr.type()).get_width();
+    if(!num_elements.has_value())
+    {
+      if(!subtype_bits.has_value() || *subtype_bits == 0)
+        num_elements = total_width;
+      else
+        num_elements = total_width / numeric_cast_v<std::size_t>(*subtype_bits);
+    }
+
+    exprt::operandst operands;
+    operands.reserve(*num_elements);
+    for(std::size_t i = 0; i < *num_elements; ++i)
+    {
+      if(subtype_bits.has_value())
+      {
+        const std::size_t subtype_bits_int =
+          numeric_cast_v<std::size_t>(*subtype_bits);
+        bitvector_typet type(bitvector_expr.type().id(), subtype_bits_int);
+        operands.push_back(bv_to_expr(
+          extractbits_exprt(
+            bitvector_expr,
+            (i + 1) * subtype_bits_int - 1,
+            i * subtype_bits_int,
+            type),
+          array_type.subtype(),
+          ns));
+      }
+      else
+      {
+        operands.push_back(
+          bv_to_expr(bitvector_expr, array_type.subtype(), ns));
+      }
+    }
+
+    return array_exprt(std::move(operands), array_type);
+  }
+  else if(target_type.id() == ID_vector)
+  {
+    const vector_typet &vector_type = to_vector_type(target_type);
+    const std::size_t num_elements =
+      numeric_cast_v<std::size_t>(vector_type.size());
+    auto subtype_bits = pointer_offset_bits(vector_type.subtype(), ns);
+
+    exprt::operandst operands;
+    operands.reserve(num_elements);
+    for(std::size_t i = 0; i < num_elements; ++i)
+    {
+      if(subtype_bits.has_value())
+      {
+        const std::size_t subtype_bits_int =
+          numeric_cast_v<std::size_t>(*subtype_bits);
+        bitvector_typet type(bitvector_expr.type().id(), subtype_bits_int);
+        operands.push_back(bv_to_expr(
+          extractbits_exprt(
+            bitvector_expr,
+            (i + 1) * subtype_bits_int - 1,
+            i * subtype_bits_int,
+            type),
+          vector_type.subtype(),
+          ns));
+      }
+      else
+      {
+        operands.push_back(
+          bv_to_expr(bitvector_expr, vector_type.subtype(), ns));
+      }
+    }
+
+    return vector_exprt(std::move(operands), vector_type);
+  }
+  else if(target_type.id() == ID_complex)
+  {
+    const complex_typet &complex_type = to_complex_type(target_type);
+    auto subtype_bits = pointer_offset_bits(complex_type.subtype(), ns);
+    std::size_t subtype_bits_int;
+    if(subtype_bits.has_value())
```

Comment:
Since this happens quite a lot, suggest clarifying in top-of-function comments: what are these two possible circumstances (known bitvector width / known target type size)? What sort of calling contexts mean that one or the other source of information is reliable? If they're both known should we be INVARIANT'ing that they're equal?

Reply:
Invariants (where possible) and documentation note added.